EiffelNet: add routines that do not raise exception

jocelyn opened the pull request:
Pull request for review.

manuseiffel commented on Src/C_library/libpng/Makefile-win.SH:
```diff
@@ -4,7 +4,7 @@ OUTDIR= .
 CC = $cc
 OUTPUT_CMD = $output_cmd
-CFLAGS = -I"$rt_include" -I..$(DIR)zlib -I.
+CFLAGS = -I"$rt_include" -I..\zlib -I.
```
Please keep $(DIR) everywhere. This is used to compile C code using the cygwin/mingw C compiler on windows.

jocelyn replied:
Ok not sure why I removed it ... mistake

manuseiffel commented on Src/C_library/libpng/Makefile-win.SH:
```diff
@@ -30,9 +30,5 @@ clean:
-	$(MKDIR) ..$(DIR)..$(DIR)compatible$(DIR)library$(DIR)vision2$(DIR)spec$(DIR)$(ISE_C_COMPILER)$(DIR)$(ISE_PLATFORM)$(DIR)lib
-	$(MV) $@ ..$(DIR)..$(DIR)compatible$(DIR)library$(DIR)vision2$(DIR)spec$(DIR)$(ISE_C_COMPILER)$(DIR)$(ISE_PLATFORM)$(DIR)lib$(DIR)$@
+	$(MKDIR) ..$(DIR)spec$(DIR)$(ISE_C_COMPILER)$(DIR)$(ISE_PLATFORM)$(DIR)lib
+	$(MV) $@ ..$(DIR)spec$(DIR)$(ISE_C_COMPILER)$(DIR)$(ISE_PLATFORM)$(DIR)lib$(DIR)$@
```
missing newline

manuseiffel commented on Src/C_library/libpng/Makefile-win.SH:
```diff
@@ -30,9 +30,5 @@ clean:
-	$(MV) $@ ..$(DIR)..$(DIR)experimental$(DIR)library$(DIR)vision2$(DIR)spec$(DIR)$(ISE_C_COMPILER)$(DIR)$(ISE_PLATFORM)$(DIR)lib$(DIR)$@
-	$(MKDIR) ..$(DIR)..$(DIR)compatible$(DIR)library$(DIR)vision2$(DIR)spec$(DIR)$(ISE_C_COMPILER)$(DIR)$(ISE_PLATFORM)$(DIR)lib
-	$(MV) $@ ..$(DIR)..$(DIR)compatible$(DIR)library$(DIR)vision2$(DIR)spec$(DIR)$(ISE_C_COMPILER)$(DIR)$(ISE_PLATFORM)$(DIR)lib$(DIR)$@
+	$(MKDIR) ..$(DIR)spec$(DIR)$(ISE_C_COMPILER)$(DIR)$(ISE_PLATFORM)$(DIR)lib
```
Wouldn't it be better to put the spec directory under the current directory, i.e. ```C_library\libpng\spec\.....```

jocelyn replied:
I don't mind either way.

manuseiffel commented on Src/C_library/libpng/Makefile.SH:
```diff
@@ -0,0 +1,23 @@
+case $CONFIG in
```
Do we actually need it? It is only for Windows.

jocelyn replied:
It makes things easier for iron

manuseiffel commented on Src/C_library/zlib/Makefile.SH:
```diff
@@ -0,0 +1,23 @@
+case $CONFIG in
```
Same as for libpng, do we need it?

manuseiffel commented on Src/library/vision2/vision2-safe.ecf:
```diff
@@ -139,6 +130,16 @@
 		</external_linker_flag>
 		<library name="base" location="$ISE_LIBRARY\library\base\base-safe.ecf"/>
+		<library name="c_library_libpng" location="$ISE_LIBRARY\C_library\libpng\clibs_libpng-safe.ecf">
+			<condition>
```
I wonder if we need the condition here, since the external is already protected?

jocelyn replied:
This is mainly to avoid loading ecf for nothing.
It will also help to have better dependency knowledge for a target.
on non windows platform, there is no dep on the c_library package.

jocelyn commented:
Merge by rev#98514

jocelyn committed:
Added socket routines that do not raise exception.

jocelyn committed:
Fixed a warning related to `len` local variable.

jocelyn committed:
Use new names for no exception routines.

jocelyn committed:
Reorganized added features.

jocelyn committed:
put same kind of routine together.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -824,81 +824,188 @@ void c_put_stream(EIF_INTEGER fd, EIF_POINTER stream_pointer, EIF_INTEGER length
+/* read operations                                    */
+/******************************************************/
+
+EIF_INTEGER c_recv_noexception(EIF_INTEGER fd, EIF_POINTER buf, EIF_INTEGER len, EIF_INTEGER flags)
```
I would remove this routine, and keep the CREADX macro but simplify it to not perform the call to eif_net_check.

jocelyn replied:
I need this c_recv_noexception.

And if we change CREADX to not raise exception I fear we may break existing user code.

manuseiffel replied:
The macro is used locally and actually you removed its usafe but not the declaration. The macros saves you a call overhead, so I would use it. If c_recv_noexception is to be used from Eiffel code, you can always, use the macro inside.

jocelyn replied:
this is what I did indeed.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -824,81 +824,188 @@ void c_put_stream(EIF_INTEGER fd, EIF_POINTER stream_pointer, EIF_INTEGER length
 {
 	float f=0.0;
-	CREADX(*l_bytes,fd,&f,sizeof(float), 0)
+	*l_bytes = c_recv_noexception(fd, &f, sizeof(float), 0);
```
Here use the CREADX macro

jocelyn replied:
Ok I think I understand why you suggested to use CREADX .. to minimize the changes.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -824,81 +824,188 @@ void c_put_stream(EIF_INTEGER fd, EIF_POINTER stream_pointer, EIF_INTEGER length
-EIF_INTEGER c_rcv_from(EIF_INTEGER fd, EIF_POINTER buf, EIF_INTEGER len, EIF_INTEGER flags, EIF_POINTER addr, EIF_POINTER addr_len)
-	/*x like c_receive and sender address is stored into socket address
-	    structure addr and address length into *addr_len */
+EIF_INTEGER c_recvfrom_noexception(EIF_INTEGER fd, EIF_POINTER buf, EIF_INTEGER len, EIF_INTEGER flags, EIF_POINTER addr, EIF_POINTER addr_len)
```
I know the existing name is bad but we should be consistent: c_rcv_from_noexception

jocelyn replied:
Could we introduce a macro c_rcv_from , and a new feature c_recvfrom  instead ?
when interfacing with C , it is much nicer to guess quickly the underlying C call...

manuseiffel replied:
This is a breaking change, so we still need it to be an exported routine if it was exported before.

jocelyn replied:
ok

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -992,6 +1114,90 @@ EIF_INTEGER c_set_sock_opt_linger(EIF_INTEGER fd, EIF_BOOLEAN flag, EIF_INTEGER
+/* Socket recv and send timeout options 				*/
+/********************************************************/
+
+EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
```
Any reasons to not have them under misc?

jocelyn replied:
It was to gather all related external together .. at first I added them to Misc 
because I was worry about platform other than Windows and Unices.

manuseiffel commented on Src/library/net/abstract/socket/socket.e:
```diff
@@ -514,6 +514,20 @@ feature -- Input
 			end
 		end
 
+	read_character_noexception
```
At the Eiffel level, I'm not sure we want those in the EiffelNet classes yet. Maybe you can add them in a SOCKET descendant for EiffelWeb. Still put the new externals in the SOCKET class.

jocelyn replied:
Ok, fine with me, I just moved them to EiffelNet for the diff ...
So I will move them back to EiffelWeb.

manuseiffel commented on Src/library/net/abstract/socket/socket.e:
```diff
@@ -1041,42 +1126,6 @@ feature {NONE} -- Implementation
 		deferred
 		end
 
-feature -- Socket Recv and Send timeout.
```
Why moving them to the descendant?

jocelyn replied:
To be coherent with other similar routines that are in the descendant ...
such as the `timeout` 
But I don't mind keeping them in SOCKET

manuseiffel replied:
If the other routines where in descendant, was it because they do not apply let's say for UNIX_SOCKET for example?

jocelyn replied:
to complete my reply ... all the timeout, blocking behavior and so on, are in NETWORK_SOCKET.
Since recv and send timeout makes sense when we have blocking behavior, I added the related feature in same class.

I could move the c externals to SOCKET.
But then, why leave the c_close, c_select_poll_with_timeout, ... in NETWORK_SOCKET.

So I prefer to try to stay coherent with EiffelNet design, even if this one could be cleaned up and improved...

manuseiffel commented on Src/library/net/network/socket/network_stream_socket.e:
```diff
@@ -79,30 +81,30 @@ feature -- Initialization
 			end
 		end
 
+	make_server_by_address_and_port (a_address: INET_ADDRESS; a_port: INTEGER)
```
Have you tested that it does work when a machine as multiple IPs?

jocelyn replied:
No explicitly but I can try, which issue do you suspect?
This will help me for the test.

manuseiffel replied:
Making sure that if you see listen to IP 10.0.10.1, an incoming connection on 10.0.10.2 would not be accepted.

jocelyn replied:
Ok I will try that soon (next week).

jocelyn commented:
A global question about your comments:

Why do you want to change CREADX ? I fear it could break some code.
Or maybe we could a new macro  CREADXNE  for No Exception ... and use it.
However here, I don't see the benefit to use a macro.
The code is not that long, and this is a direct call to the C functions.

jocelyn committed:
Removed Eiffel feature for noexception, and leave only the C external for now.
Added noexception C features wrapping `send'.
Marked as Unsed a few C features.

jocelyn reviewed (commented):
Is this LGTM now ?

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -726,26 +726,36 @@ EIF_INTEGER c_peer_name(EIF_INTEGER s, EIF_POINTER addr, EIF_INTEGER length)
 void c_send_char_to(EIF_INTEGER fd, EIF_CHARACTER c, EIF_INTEGER flags, EIF_POINTER addr_pointer, EIF_INTEGER sizeofaddr)
 	/*x transmission of character c through socket fd */
 {
+	/* Unused (!) by EiffelNet */
```
We can remove it because this is actually only used by the obsolete version of EiffelNet which has its own C code.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -726,26 +726,36 @@ EIF_INTEGER c_peer_name(EIF_INTEGER s, EIF_POINTER addr, EIF_INTEGER length)
 void c_send_int_to(EIF_INTEGER fd, EIF_INTEGER i, EIF_INTEGER flags, EIF_POINTER addr_pointer, EIF_INTEGER sizeofaddr)
 	/* transmission of Eiffel integer i through socket fd */
 {
+	/* Unused (!) by EiffelNet */
```
ditto

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -758,6 +768,7 @@ void c_send_float_to(EIF_INTEGER fd, EIF_REAL f, EIF_INTEGER flags, EIF_POINTER
 	   the caller,an automatic conversion from float to double will occur.
 	   We need to explicitly convert the argument to a float before safely
 	   sending it on the socket. Xavier */
+	/* Unused (!) by EiffelNet */
```
ditto

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -767,29 +778,34 @@ void c_send_float_to(EIF_INTEGER fd, EIF_REAL f, EIF_INTEGER flags, EIF_POINTER
 void c_send_double_to(EIF_INTEGER fd, EIF_DOUBLE d, EIF_INTEGER flags, EIF_POINTER addr_pointer, EIF_INTEGER sizeofaddr)
 	/*x transmission of double d through socket fd */
 {
+	/* Unused (!) by EiffelNet */
```
ditto

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -767,29 +778,34 @@ void c_send_float_to(EIF_INTEGER fd, EIF_REAL f, EIF_INTEGER flags, EIF_POINTER
 void c_put_char(EIF_INTEGER fd, EIF_CHARACTER c)
 	/*x transmission of character c through socket fd */
 {
+	/* Unused (!) by EiffelNet */
```
ditto

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -767,29 +778,34 @@ void c_send_float_to(EIF_INTEGER fd, EIF_REAL f, EIF_INTEGER flags, EIF_POINTER
 void c_put_int(EIF_INTEGER fd, EIF_INTEGER i)
 	/* transmission of Eiffel integer i through socket fd */
 {
+	/* Unused (!) by EiffelNet */
```
ditto

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -802,6 +818,7 @@ void c_put_float(EIF_INTEGER fd, EIF_REAL f)
 	   caller,an automatic conversion from float to double will occur.
 	   We need to explicitly convert the argument to a float before safely
 	   sending it on the socket. Xavier */
+	/* Unused (!) by EiffelNet */
```
ditto

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -812,93 +829,180 @@ void c_put_float(EIF_INTEGER fd, EIF_REAL f)
 void c_put_double(EIF_INTEGER fd, EIF_DOUBLE d)
 	/*x transmission of double d through socket fd */
 {
+	/* Unused (!) by EiffelNet */
```
ditto

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -726,26 +726,36 @@ EIF_INTEGER c_peer_name(EIF_INTEGER s, EIF_POINTER addr, EIF_INTEGER length)
+
 #define CPUTX(descriptor,element_pointer,sizeofelement) \
-	eif_net_check (send ((EIF_SOCKET_TYPE) descriptor, (char *) element_pointer, (int) sizeofelement, (int) 0));
+	eif_net_check (send((EIF_SOCKET_TYPE) descriptor, (char *) element_pointer, (int) sizeofelement, (int) 0));
```
Let's remove this CPUTX macro.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -812,93 +829,180 @@ void c_put_float(EIF_INTEGER fd, EIF_REAL f)
 void c_put_stream(EIF_INTEGER fd, EIF_POINTER stream_pointer, EIF_INTEGER length)
 	/*x transmission of string s of size size trought socket fd */
 {
 	CPUTX(fd, (char *) stream_pointer,length)
```
Use CSENDX macro instead with an explicit call to eif_net_check. That will be more clear that c_put_stream and c_put_stream_noexception have similar implementation and only differs by the call to eif_net_check. This will be similar to the new c_read_stream and c_read_stream_noexception implementations

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -812,93 +829,180 @@ void c_put_float(EIF_INTEGER fd, EIF_REAL f)
-	CREADX(result,fd,buf,len,flags);
-	return (EIF_INTEGER) result;
+	EIF_INTEGER res;
+	res = c_recv_noexception(fd, buf, len, flags);
```
Let's use the macro to be consistent.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -812,93 +829,180 @@ void c_put_float(EIF_INTEGER fd, EIF_REAL f)
 	    return number of actually sent bytes */
 {
+	EIF_INTEGER res;
+	res = c_write_noexception(fd, buf, len);
```
Same here use the macro here.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -908,22 +1012,37 @@ EIF_INTEGER c_send(EIF_INTEGER fd, EIF_POINTER buf, EIF_INTEGER len, EIF_INTEGER
 	    return actual number of bytes sent */
 {
+	EIF_INTEGER res;
+	res = c_send_noexception(fd, buf, len, flags);
```
Use the macro here.

manuseiffel reviewed (approved):
A few changes but good overall.

jocelyn commented:
committed on subversion repo by rev#99272 , rev#99273, rev#99274
